sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -48,6 +48,9 @@ object ConstantFolding extends Rule[LogicalPlan] {
       // object and running eval unnecessarily.
       case l: Literal => l
 
+      case Size(c: CreateArray, _) => Literal(c.children.length)
+      case Size(c: CreateMap, _) => Literal(c.children.length / 2)
```

Comment:
Hm .. one case I can imagine is the case when the children of `CreateArray` or `CreateMap` throw an exception e.g.) from a UDF or when ANSI mode is enabled. After this change, looks it's going to suppress such exceptions. For example:

```scala
spark.range(1).selectExpr("size(array(assert_true(false)))").show()
```

As an alternative, we could maybe only allow this when the children of `CreateArray` or `CreateMap` are safe and don't throw an exception? e.g.) only when they are `AttributeReference`s?

Reply:
Good point. This is a common issue when trying to optimize expression tree.

Reply:
good point. We should only allow attributes and literals.

Reply:
so the problem with this is it doesn't work with nested.. like inline with array of struct:

f(CreateArray(Seq(CreateStruct(Seq('c1)))))

Unless we want to recurse down. I don't think there is any one type we can use for the complex type generators though, thoughts?

Reply:
Seems we need a way to propagate the no-side-effect property:
```
def hasNoSideEffect(e: Expression): Boolean = e match {
  case _: Attribute => true
  case _: Literal => true
  case _: NoSideEffect => e.children.forall(hasNoSideEffect)
  case _ => false
}
...
case Size(c: CreateArray, _) if c.children.forall(hasNoSideEffect) => Literal(c.children.length)
case Size(c: CreateMap, _) if c.children.forall(hasNoSideEffect) => Literal(c.children.length / 2)
```

`CreateStruct/Array/Map` can extend the trait `NoSideEffect`, and we can add more in the future.